### PR TITLE
Added a timer tool

### PR DESCRIPTION
**Pull request by zacharyselk**
A tool added to the RARS GUI that simulates a timing core, accessed via MMIO, that can handle timing interrupts as referenced in #3. It should (hopefully) be entirely compliant to the RISC-V spec aside from handling privilege levels (ie it does not check for user, supervisor, and machine level permissions).

(Also this is my first time attempting to code in Java so the code might look a bit odd / might be implemented in an odd way)

### CODE REVIEW

File: rars/tools/TimerTool.java [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+                int address = notice.getAddress();
+                int value = notice.getValue();
+
+                // Check what word was changed, then update the corrisponding information
+                if (address == TIME_CMP_ADDRESS) {
+                    this.value = ((this.value >> 32) << 32) + value;
+                    postInterrupt = true; // timecmp was writen to
+                }
+                else if (address == TIME_CMP_ADDRESS+4) {
+                    this.value = (this.value & 0xFFFFFFFF) + (value << 32);
```
Comment (TheThirdOne) [inline rars/tools/TimerTool.java]: `value << 32`  should be always 0. Did you test setting the upper 4 bytes?

File: rars/tools/TimerTool.java [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+    private static long startTime = 0L; // Tmp unix time used to keep track of how much time has passed
+    private static long savedTime = 0L; // Accumulates time as we pause/play the timer
+
+    // Timing threads
+    private static TimeCmpDaemon timeCmp = null; // Watches for changes made to timecmp
+    private Timer timer = new Timer();
+    private Tick tick = new Tick(); // Runs every millisecond to decide if a timer inturrupt should be raised
+
+    // Internal timing flags
+    private static boolean postInterrupt = false; // Signals when timecmp has been writen to
```
Comment (TheThirdOne) [inline rars/tools/TimerTool.java]: This line appears to be unused whereas the one in TimeCmpDaemon is used. Did you forget to delete this one?

**Comment by TheThirdOne** [comment]
From a quick review, it looks pretty good. The main issue I see is that build/ and rars.jar were committed. I should have put them in the gitignore when I made the build script. I'll removed the references when I merge this in. 

Additionally, it would be nice to have an example program that uses the tool. If you don't already have a program you were testing with, I can make one without much trouble; your help page should be sufficient documentation.

File: rars/tools/TimerTool.java (hunk 1 above)
Comment (zacharyselk) [reply rars/tools/TimerTool.java]: I totally forgot to test out times larger than 32-bits, I will get on that.

File: rars/tools/TimerTool.java (hunk 2 above)
Comment (zacharyselk) [reply rars/tools/TimerTool.java]: Yes I did.

**Comment by zacharyselk** [comment]
Great to hear! I added a small example program, timerTest.s, that uses the tool.

**Comment by TheThirdOne** [comment]
This was merged in with 5e5f218e296c62e80df5915fb3fa7fcd7418e12b.

Thanks for the PR. The interrupt side of RARS definitely needed some attention and this is a good step towards improving it.

**Comment by colbac8** [comment]
I'm trying to use the timer tool starting from the test file timertool.s, but I've some questions about. 
When I start the timertool.s program, it seems to remain in the loop forever. No timing interrupts are generated after 5 seconds (i.e., the label "hello" is never printed.). Do I have to enable something in RARS or do something else to make the timertool.s program work?

**Comment by zacharyselk** [comment]
Yes, in RARS to use timer interrupts you need to enable and run the Timer Tool. This is done by going to `Tools > Timer Tool` and then clicking the `Connect to Program` button then clicking the `Play` button.

**Comment by colbac8** [comment]
Yes, I tried by using the timer tool but it didn't work in the version rars_1_3_1. I download the latest version (1.4) and it now works! Thank you

**Pull request closed without merging**